src/pinnable_mapped_file.cpp

```diff
@@ -12,48 +12,99 @@
+      case db_error_code::locked_mode_required:
+	 return "Locked mode is required for hugepage usage";
+      case db_error_code::not_found:
+	 return "Database file not found";
+      case db_error_code::bad_size:
+	 return "Bad size";
+      case db_error_code::no_huge_page:
+	 return "Hugepage support is a linux-only feature";
+      case db_error_code::no_locked_mode:
+	 return "Locked mode not supported on win32";
+      case db_error_code::bad_header:
+	 return "Failed to read DB header";
+      case db_error_code::no_access:
+	 return "Could not gain write access to the shared memory file";
+      case db_error_code::aborted:
+	 return "Database load aborted";
+      case db_error_code::no_mlock:
+	 return "Failed to mlock database";
+      default:
+         return "Unrecognized error code";
+   }
+}
+
+const std::error_category& chainbase_error_category() {
+   static class chainbase_error_category the_category;
+   return the_category;
+}
+
 pinnable_mapped_file::pinnable_mapped_file(const bfs::path& dir, bool writable, uint64_t shared_file_size, bool allow_dirty,
                                           map_mode mode, std::vector<std::string> hugepage_paths) :
    _data_file_path(bfs::absolute(dir/"shared_memory.bin")),
    _database_name(dir.filename().string()),
    _writable(writable)
 {
-   if(shared_file_size % _db_size_multiple_requirement)
-      BOOST_THROW_EXCEPTION(std::runtime_error("Database must be mulitple of " + std::to_string(_db_size_multiple_requirement) + " bytes"));
+   if(shared_file_size % _db_size_multiple_requirement) {
+      std::cerr << "Database must be mulitple of " <<  std::to_string(_db_size_multiple_requirement) << " bytes" << std::endl;
+      BOOST_THROW_EXCEPTION(std::system_error(make_error_code(db_error_code::bad_size)));
+   }
 #ifndef __linux__
    if(hugepage_paths.size())
-      BOOST_THROW_EXCEPTION(std::runtime_error("Hugepage support is a linux only feature"));
+      BOOST_THROW_EXCEPTION(std::system_error(make_error_code(db_error_code::no_huge_page)));
 #endif
    if(hugepage_paths.size() && mode != locked)
-      BOOST_THROW_EXCEPTION(std::runtime_error("Locked mode is required for hugepage usage"));
+      BOOST_THROW_EXCEPTION(std::system_error(make_error_code(db_error_code::locked_mode_required)));
 #ifdef _WIN32
    if(mode == locked)
-      BOOST_THROW_EXCEPTION(std::runtime_error("Locked mode not supported on win32"));
+      BOOST_THROW_EXCEPTION(std::system_error(make_error_code(db_error_code::no_locked_mode)));
 #endif
 
-   if(!_writable && !bfs::exists(_data_file_path))
-      BOOST_THROW_EXCEPTION(std::runtime_error("database file not found at " + _data_file_path.string()));
+   if(!_writable && !bfs::exists(_data_file_path)){
+      std::cerr << "database file not found at " <<  _data_file_path.string() << std::endl;
+      BOOST_THROW_EXCEPTION(std::system_error(make_error_code(db_error_code::not_found)));
```

Comment:
Instead of `cerr` create a string with this info and pass it into the `system_error` constructor